File: Shell~/HadesControl/Sources/HadesControl/ControlClient.swift
import Foundation

/// Every way a control-API call can fail to hand back a decoded DTO. `staleToken` is deliberately
/// its own case, not folded into `.server(status: 401, ...)`: a 401 here means specifically that
/// the token this client was built with is stale, almost always because the core restarted and
/// wrote a fresh discovery file - the correct, and only, recovery is to call `Discovery.read()`
/// again and build a new `ControlClient`. Retrying the same request with the same token will fail
/// the same way every time. Making this a distinct case (rather than a status code the caller has
/// to compare) is what makes it actionable in the type, per this package's own contract.
public enum ControlClientError: Error, Sendable {
    /// The token this client presented was rejected (HTTP 401). Re-read the discovery file and
    /// build a new `ControlClient`.
    case staleToken

    /// A non-2xx, non-401 response. `message` is the server's own `"error"` field when the body
    /// carried one (every Control endpoint's error responses do - see `ControlAuth`,
    /// `ProjectsEndpoint`, `EditorsEndpoint`), never text invented client-side.
    case server(status: Int, message: String?)

    /// The response body did not decode into the expected DTO shape.
    case decoding(DecodingError)

    /// A request body (e.g. `AddProjectRequest`, `WriteMemoryDocumentRequest`) failed to encode to
    /// JSON. Every request body this package actually sends is a plain-String-fields struct, which
    /// `JSONEncoder` cannot fail to encode - this case exists only so the client stays total under
    /// Swift's typed-throws checking, the same reasoning `send`'s own decode fallback documents.
    case encoding(EncodingError)

    /// The request never got a response to check the status of - the core is not running, this
    /// connection's port is stale, the request timed out, and so on.
    case transport(URLError)
}

/// The control-API client: every phase-one endpoint, decoded straight into the DTOs in
/// `DTOs.swift` and nothing else. No retry policy, no caching, no derived state - a thin,
/// stateless wrapper over `URLSession` that attaches the bearer token and maps the response to
/// either a DTO or a `ControlClientError`.
public struct ControlClient: Sendable {
    private let baseURL: URL
    private let token: String
    private let session: URLSession

    /// `connection` is normally the result of `Discovery.read()`. `session` is overridable so
    /// tests can substitute a mocked `URLProtocol` instead of making real network calls.
    public init(connection: ControlConnection, session: URLSession = .shared) {
        // 127.0.0.1, never "localhost": matches ControlListener's own loopback bind exactly, and
        // sidesteps a DNS resolution step for what is always a same-machine call.
        self.baseURL = URL(string: "http://127.0.0.1:\(connection.port)")!
        self.token = connection.token
        self.session = session
    }

    /// `GET /control/ping`.
    public func ping() async throws(ControlClientError) -> PingResult {
        try await get("/control/ping")
    }

    /// `GET /control/summary` - the menu bar's one endpoint.
    public func summary() async throws(ControlClientError) -> SummaryResult {
        try await get("/control/summary")
    }

    /// `GET /control/projects`.
    public func projects() async throws(ControlClientError) -> ProjectsResult {
        try await get("/control/projects")
    }

    /// `GET /control/editors`.
    public func editors() async throws(ControlClientError) -> EditorsResult {
        try await get("/control/editors")
    }

    /// `POST /control/leases/{id}/release`. `id` is the holding project's `productGuid` (see
    /// `SummaryLease.leaseId`'s own doc comment) - percent-encoded defensively, matching the
    /// `hades` CLI's own `Uri.EscapeDataString`, even though a productGuid is always plain hex in
    /// practice. Idempotent and safe to call late: if the TTL already released the lease
    /// server-side, this still answers `success: true` with an explanatory `message`, never an
    /// error - see `ActionResult`'s own doc comment. Callers must not synthesize an error for a
    /// `success: false` result either; `message` already names what happened.
    public func releaseLease(id: String) async throws(ControlClientError) -> ActionResult {
        try await post("/control/leases/\(encodedPathSegment(id))/release")
    }

    // MARK: - Projects actions

    /// `POST /control/projects/add`. The panel that chose `path` is the only place the shell picks
    /// one - this call just adopts and fully indexes it, returning the same `ProjectRow` shape
    /// `projects()` uses for every other row.
    public func addProject(path: String) async throws(ControlClientError) -> ProjectRow {
        try await post("/control/projects/add", body: AddProjectRequest(path: path))
    }

    /// `POST /control/projects/{productGuid}/remove`. Deregisters only - see `ActionResult.message`
    /// (and this endpoint's own .NET doc comment) for why nothing on disk is ever deleted by this
    /// call, including the project's own graph and authored memory.
    public func removeProject(productGuid: String) async throws(ControlClientError) -> ActionResult {
        try await post("/control/projects/\(encodedPathSegment(productGuid))/remove")
    }

    /// `POST /control/projects/{productGuid}/rebuild`. Returns immediately with an operation id
    /// pollable via `operation(id:)` from the moment this call returns.
    public func rebuildProject(productGuid: String) async throws(ControlClientError) -> RebuildStartedResult {
        try await post("/control/projects/\(encodedPathSegment(productGuid))/rebuild")
    }

    /// `POST /control/projects/{productGuid}/installPlugin`.
    public func installPlugin(productGuid: String) async throws(ControlClientError) -> InstallPluginResult {
        try await post("/control/projects/\(encodedPathSegment(productGuid))/installPlugin")
    }

    /// `POST /control/projects/{productGuid}/revealInFinder`.
    public func revealInFinder(productGuid: String) async throws(ControlClientError) -> ActionResult {
        try await post("/control/projects/\(encodedPathSegment(productGuid))/revealInFinder")
    }

    /// `POST /control/projects/{productGuid}/openInUnity`.
    public func openInUnity(productGuid: String) async throws(ControlClientError) -> ActionResult {
        try await post("/control/projects/\(encodedPathSegment(productGuid))/openInUnity")
    }

    // MARK: - Operations

    /// `GET /control/operations/{id}` - the poll side of `rebuildProject(productGuid:)` (today's
    /// only long-running action). An unknown id is a `.server(status: 404, ...)` naming exactly why
    /// (it may have completed and been pruned, or the id is wrong) - never treat that as a generic
    /// failure the caller must guess at.
    public func operation(id: String) async throws(ControlClientError) -> OperationResult {
        try await get("/control/operations/\(encodedPathSegment(id))")
    }

    // MARK: - Settings

    /// `GET /control/settings`.
    public func settings() async throws(ControlClientError) -> SettingsResult {
        try await get("/control/settings")
    }

    // MARK: - Traces

    /// `GET /control/traces/sequences`. `project` is a flexible handle (name or productGuid,
    /// matching `hades_status`'s own "project" argument) - omit it when only one project is known.
    /// Every other filter is applied AFTER the core groups calls into sequences, never before (see
    /// `Hades.Server.Control.TracesEndpoint`'s own doc comment for why filtering first would
    /// corrupt grouping) - Swift does not re-filter client-side either. `limit` is omitted (not
    /// defaulted to today's `200`) when nil, so the core's own route default stays the one source
    /// of truth for it - Swift never keeps a stale copy of a server-owned policy value.
    public func tracesSequences(
        project: String? = nil, tool: String? = nil, outcome: String? = nil,
        minDurationMs: Int? = nil, maxDurationMs: Int? = nil, limit: Int? = nil
    ) async throws(ControlClientError) -> TraceSequencesResult {
        var query: [URLQueryItem] = []
        if let project { query.append(URLQueryItem(name: "project", value: project)) }
        if let tool { query.append(URLQueryItem(name: "tool", value: tool)) }
        if let outcome { query.append(URLQueryItem(name: "outcome", value: outcome)) }
        if let minDurationMs { query.append(URLQueryItem(name: "minDurationMs", value: String(minDurationMs))) }
        if let maxDurationMs { query.append(URLQueryItem(name: "maxDurationMs", value: String(maxDurationMs))) }
        if let limit { query.append(URLQueryItem(name: "limit", value: String(limit))) }
        return try await get("/control/traces/sequences", query: query)
    }

    /// `GET /control/traces/{traceId}` - one trace's full span detail.
    public func traceDetail(traceId: String, project: String? = nil) async throws(ControlClientError) -> TraceDetailResult {
        var query: [URLQueryItem] = []
        if let project { query.append(URLQueryItem(name: "project", value: project)) }
        return try await get("/control/traces/\(encodedPathSegment(traceId))", query: query)
    }

    /// `GET /control/traces/slow` - the slowest tools, ranked, surfaced from their own endpoint
    /// rather than sorted client-side out of `tracesSequences`. `limit` is omitted when nil - see
    /// `tracesSequences`'s own doc comment for why.
    public func tracesSlow(project: String? = nil, limit: Int? = nil) async throws(ControlClientError) -> SlowToolsResult {
        var query: [URLQueryItem] = []
        if let project { query.append(URLQueryItem(name: "project", value: project)) }
        if let limit { query.append(URLQueryItem(name: "limit", value: String(limit))) }
        return try await get("/control/traces/slow", query: query)
    }

    /// `GET /control/traces/failures` - failed calls, surfaced from their own endpoint rather than
    /// filtered client-side out of `tracesSequences`. `limit` is omitted when nil - see
    /// `tracesSequences`'s own doc comment for why.
    public func tracesFailures(project: String? = nil, limit: Int? = nil) async throws(ControlClientError) -> FailedCallsResult {
        var query: [URLQueryItem] = []
        if let project { query.append(URLQueryItem(name: "project", value: project)) }
        if let limit { query.append(URLQueryItem(name: "limit", value: String(limit))) }
        return try await get("/control/traces/failures", query: query)
    }

    // MARK: - Memory

    /// `GET /control/memory` - every authored document AND the proposal queue, in one round trip.
    public func memory(project: String? = nil) async throws(ControlClientError) -> MemoryResult {
        var query: [URLQueryItem] = []
        if let project { query.append(URLQueryItem(name: "project", value: project)) }
        return try await get("/control/memory", query: query)
    }

    /// `GET /control/memory/document` - one document's complete raw text. `name` is validated
    /// server-side as a basename (no traversal, no rooted paths) - see the Task 1 report for
    /// confirmation this is actually enforced, not just documented.
    public func memoryDocument(name: String, project: String? = nil) async throws(ControlClientError) -> MemoryDocumentResult {
        var query: [URLQueryItem] = []
        if let project { query.append(URLQueryItem(name: "project", value: project)) }
        query.append(URLQueryItem(name: "name", value: name))
        return try await get("/control/memory/document", query: query)
    }

    /// `POST /control/memory/document` - writes (creating or overwriting) one authored document
    /// verbatim. This is the shell's own text-editor save path, not a merge - see
    /// `acceptMemoryProposal(fileName:project:)` for the one path that appends instead.
    public func writeMemoryDocument(
        name: String, content: String, project: String? = nil
    ) async throws(ControlClientError) -> ActionResult {
        var query: [URLQueryItem] = []
        if let project { query.append(URLQueryItem(name: "project", value: project)) }
        query.append(URLQueryItem(name: "name", value: name))
        return try await post("/control/memory/document", query: query, body: WriteMemoryDocumentRequest(content: content))
    }

    /// `POST /control/memory/proposals/accept` - appends the proposal's content into its own
    /// `targetFile` (creating it if needed) and marks the proposal accepted; never deletes the
    /// proposal file itself.
    public func acceptMemoryProposal(fileName: String, project: String? = nil) async throws(ControlClientError) -> ActionResult {
        var query: [URLQueryItem] = []
        if let project { query.append(URLQueryItem(name: "project", value: project)) }
        query.append(URLQueryItem(name: "fileName", value: fileName))
        return try await post("/control/memory/proposals/accept", query: query)
    }

    /// `POST /control/memory/proposals/dismiss` - deletes the proposal file. `confirm` must be
    /// explicitly `true`; the core refuses with a 400 otherwise rather than silently no-op'ing or
    /// deleting anyway - the shell's own confirmation UI is what should set this to `true`, never a
    /// default.
    public func dismissMemoryProposal(
        fileName: String, confirm: Bool, project: String? = nil
    ) async throws(ControlClientError) -> ActionResult {
        var query: [URLQueryItem] = []
        if let project { query.append(URLQueryItem(name: "project", value: project)) }
        query.append(URLQueryItem(name: "fileName", value: fileName))
        query.append(URLQueryItem(name: "confirm", value: confirm ? "true" : "false"))
        return try await post("/control/memory/proposals/dismiss", query: query)
    }

    /// `POST /control/memory/proposals/defer` - pure bookkeeping: marks the proposal deferred,
    /// never deletes it, never writes an authored document.
    public func deferMemoryProposal(fileName: String, project: String? = nil) async throws(ControlClientError) -> ActionResult {
        var query: [URLQueryItem] = []
        if let project { query.append(URLQueryItem(name: "project", value: project)) }
        query.append(URLQueryItem(name: "fileName", value: fileName))
        return try await post("/control/memory/proposals/defer", query: query)
    }

    // MARK: - Request plumbing

    private func get<Response: Decodable>(_ path: String, query: [URLQueryItem] = []) async throws(ControlClientError) -> Response {
        try await send(request(method: "GET", path: path, query: query))
    }

    private func post<Response: Decodable>(_ path: String, query: [URLQueryItem] = []) async throws(ControlClientError) -> Response {
        try await send(request(method: "POST", path: path, query: query))
    }

    /// The one call site every request-body endpoint (`addProject`, `writeMemoryDocument`) goes
    /// through - `query` and `body` compose independently, so a route needing both (`writeMemoryDocument`'s
    /// `name`/`project` query items alongside its JSON `content` body) is one call, not two request
    /// objects merged after the fact.
    private func post<Body: Encodable, Response: Decodable>(
        _ path: String, query: [URLQueryItem] = [], body: Body
    ) async throws(ControlClientError) -> Response {
        var httpRequest = request(method: "POST", path: path, query: query)
        do {
            httpRequest.httpBody = try JSONEncoder().encode(body)
            httpRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        } catch let error as EncodingError {
            throw .encoding(error)
        } catch {
            // JSONEncoder's documented failure mode is EncodingError; this branch only exists so
            // this function stays total under Swift's typed-throws checking - see `send`'s own
            // matching branch for the same reasoning against JSONDecoder.
            throw .encoding(.invalidValue(body, .init(codingPath: [], debugDescription: "\(error)")))
        }
        return try await send(httpRequest)
    }

    /// Percent-encodes one caller-supplied path segment (a productGuid, operation id, or traceId) -
    /// defensive even though every one of these is plain hex or a GUID in practice, matching
    /// `releaseLease(id:)`'s own existing precedent (see that method's own doc comment).
    private func encodedPathSegment(_ raw: String) -> String {
        raw.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? raw
    }

    private func request(method: String, path: String, query: [URLQueryItem] = []) -> URLRequest {
        // Every path passed in above is a literal control-API route this file owns; the one
        // caller-supplied path SEGMENT (a lease/operation id, a productGuid, a traceId) is already
        // percent-encoded before it reaches here - see `releaseLease` and `encodedPathSegment`.
        var components = URLComponents(url: URL(string: baseURL.absoluteString + path)!, resolvingAgainstBaseURL: false)!
        if !query.isEmpty { components.queryItems = query }

        var request = URLRequest(url: components.url!)
        request.httpMethod = method
        // Every request carries the bearer token - reads as well as writes, matching
        // ControlAuth.UseControlTokenAuth's own "applied globally, before any endpoint is mapped"
        // contract on the server side.
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        return request
    }

    private func send<Response: Decodable>(_ request: URLRequest) async throws(ControlClientError) -> Response {
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch let error as URLError {
            throw .transport(error)
        } catch {
            // URLSession's documented failure mode for `data(for:)` is URLError; anything else
            // (e.g. a test double throwing a different Error) still needs to surface as a
            // transport failure rather than crash.
            throw .transport(URLError(.unknown))
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw .transport(URLError(.badServerResponse))
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            if httpResponse.statusCode == 401 {
                throw .staleToken
            }
            let message = try? JSONDecoder().decode(ControlErrorBody.self, from: data).error
            throw .server(status: httpResponse.statusCode, message: message)
        }

        do {
            return try JSONDecoder().decode(Response.self, from: data)
        } catch let error as DecodingError {
            throw .decoding(error)
        } catch {
            // JSONDecoder's documented failure mode is DecodingError; this branch only exists so
            // the function stays total under Swift's typed-throws checking.
            throw .decoding(.dataCorrupted(.init(codingPath: [], debugDescription: "\(error)")))
        }
    }
}

/// The `{"error": "..."}` shape every non-2xx Control response body carries - see
/// `ControlAuth.UseControlTokenAuth`, `ProjectsEndpoint.Remove`, `EditorsEndpoint.ReleaseAsync`,
/// and every other error path in the control API. Not a public DTO: it exists only so
/// `ControlClient` can read the server's own message for `ControlClientError.server(message:)`
/// rather than inventing one.
private struct ControlErrorBody: Decodable {
    let error: String
}

File: Core/src/Hades.Control.Client/ControlClient.cs
using System.Globalization;
using System.Net;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Serialization;
using Hades.Control.Client.Dtos;

namespace Hades.Control.Client;

/// <summary>
/// The control-API client: every read-only endpoint that exists so far, decoded straight into the
/// DTOs in <c>Dtos/</c> and nothing else. No retry policy, no caching, no derived state - a thin,
/// stateless wrapper over <see cref="HttpClient"/> that attaches the bearer token once and maps
/// every response to either a DTO or a <see cref="ControlClientException"/>.
///
/// The .NET twin of Swift's <c>ControlClient</c>
/// (Mac/HadesControl/Sources/HadesControl/ControlClient.swift) - see that file's own class doc
/// comment for the "renders what the core decided and nothing else" contract this port holds
/// itself to as well. Beyond the phase-one read routes (ping, summary, projects, settings,
/// editors) and the one mutation <c>Hades.Cli</c> needs (release a lease), every other
/// action/mutation route (add/remove/rebuild project, edit memory, migration) that the Swift
/// client already speaks is deliberately left for a later task - this file does not guess at, or
/// stub out, routes it has not yet confirmed a caller for.
/// </summary>
public sealed class ControlClient
{
    readonly HttpClient _http;

    /// <summary><paramref name="connection"/> is normally the result of <c>Discovery.Read</c>.
    /// <paramref name="httpClient"/> is overridable so tests can substitute a stub
    /// <see cref="HttpMessageHandler"/> instead of making real network calls - same reasoning as
    /// the Swift original's own <c>session: URLSession</c> constructor parameter.</summary>
    public ControlClient(ControlConnection connection, HttpClient? httpClient = null)
    {
        _http = httpClient ?? new HttpClient();

        // 127.0.0.1, never "localhost": matches ControlListener's own loopback bind exactly, and
        // sidesteps a DNS resolution step for what is always a same-machine call - see the Swift
        // original's identical reasoning on its own baseURL.
        _http.BaseAddress = new Uri($"http://127.0.0.1:{connection.Port}");

        // Every request carries the bearer token - reads as well as writes, matching
        // ControlAuth.UseControlTokenAuth's own "applied globally, before any endpoint is mapped"
        // contract on the server side. Set once here, not per-request, since it never changes for
        // the lifetime of a client built over one ControlConnection.
        _http.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", connection.Token);
    }

    /// <summary><c>GET /control/ping</c>.</summary>
    public Task<PingResult> PingAsync() => GetAsync<PingResult>("/control/ping");

    /// <summary><c>GET /control/summary</c> - the menu bar's one endpoint.</summary>
    public Task<SummaryResult> SummaryAsync() => GetAsync<SummaryResult>("/control/summary");

    /// <summary><c>GET /control/projects</c>.</summary>
    public Task<ProjectsResult> ProjectsAsync() => GetAsync<ProjectsResult>("/control/projects");

    /// <summary><c>GET /control/settings</c>.</summary>
    public Task<SettingsResult> SettingsAsync() => GetAsync<SettingsResult>("/control/settings");

    /// <summary><c>GET /control/editors</c>.</summary>
    public Task<EditorsResult> EditorsAsync() => GetAsync<EditorsResult>("/control/editors");

    /// <summary><c>POST /control/leases/{id}/release</c> - force-releases the reload lease
    /// <paramref name="leaseId"/> names (the same id <see cref="SummaryLease.LeaseId"/> reports).
    /// Idempotent on the server side: releasing when nothing is held still succeeds.</summary>
    public Task<ActionResult> ReleaseLeaseAsync(string leaseId) =>
        PostAsync<ActionResult>($"/control/leases/{Uri.EscapeDataString(leaseId)}/release");

    /// <summary><c>GET /control/operations/{id}</c> - one tracked operation's current state.
    /// An unknown id answers 404 with the server's own explanation ("It may have completed and been
    /// pruned, or the id is wrong"), which is an ORDINARY outcome for a rebuild that finished a
    /// while ago rather than a failure; callers tell it apart by
    /// <see cref="ControlClientException.StatusCode"/>.</summary>
    public Task<OperationResult> OperationAsync(string id) =>
        GetAsync<OperationResult>($"/control/operations/{Uri.EscapeDataString(id)}");

    /// <summary><c>POST /control/projects/add</c>. Answers the new <see cref="ProjectRow"/> itself -
    /// deliberately no <c>message</c> field, because the row appearing in the list IS the feedback,
    /// so there is no server text for a caller to display and none should be invented.</summary>
    public Task<ProjectRow> AddProjectAsync(string path) =>
        PostAsync<AddProjectRequest, ProjectRow>("/control/projects/add", new AddProjectRequest { Path = path });

    /// <summary><c>POST /control/projects/{productGuid}/remove</c>.</summary>
    public Task<ActionResult> RemoveProjectAsync(string productGuid) =>
        PostAsync<ActionResult>($"/control/projects/{Uri.EscapeDataString(productGuid)}/remove");

    /// <summary><c>POST /control/projects/{productGuid}/rebuild</c> - starts a rebuild and hands
    /// back the operation id to poll with <see cref="OperationAsync"/>. Returns as soon as the
    /// operation is registered; it does not wait for the rebuild.</summary>
    public Task<RebuildStartedResult> RebuildProjectAsync(string productGuid) =>
        PostAsync<RebuildStartedResult>($"/control/projects/{Uri.EscapeDataString(productGuid)}/rebuild");

    /// <summary><c>POST /control/projects/{productGuid}/installPlugin</c>.
    /// <see cref="InstallPluginResult.Message"/> already says whether a restart is needed in plain
    /// language - render it verbatim rather than re-stating <c>needsRestart</c> as your own text.</summary>
    public Task<InstallPluginResult> InstallPluginAsync(string productGuid) =>
        PostAsync<InstallPluginResult>($"/control/projects/{Uri.EscapeDataString(productGuid)}/installPlugin");

    /// <summary><c>POST /control/projects/{productGuid}/revealInFinder</c>. The route keeps its
    /// macOS name on both platforms - it is the server's, and the server decides what revealing
    /// means per platform (Explorer on Windows). A client that renamed it would be guessing.</summary>
    public Task<ActionResult> RevealInFinderAsync(string productGuid) =>
        PostAsync<ActionResult>($"/control/projects/{Uri.EscapeDataString(productGuid)}/revealInFinder");

    /// <summary><c>POST /control/projects/{productGuid}/openInUnity</c>.</summary>
    public Task<ActionResult> OpenInUnityAsync(string productGuid) =>
        PostAsync<ActionResult>($"/control/projects/{Uri.EscapeDataString(productGuid)}/openInUnity");

    /// <summary>
    /// <c>GET /control/traces/sequences</c> - the primary timeline, and the only traces route that
    /// accepts filters at all.
    ///
    /// Filtering happens SERVER-side: a caller must not re-filter the result, because the server
    /// groups calls into sequences before filtering and doing it the other way round corrupts the
    /// grouping. <paramref name="limit"/> is OMITTED when null rather than defaulted to today's 200,
    /// so the route's own default stays the single source of truth - a client that hardcoded it
    /// would be keeping a stale copy of a server-owned policy value.
    /// </summary>
    public Task<TraceSequencesResult> TracesSequencesAsync(
        string? project = null,
        string? tool = null,
        string? outcome = null,
        long? minDurationMs = null,
        long? maxDurationMs = null,
        int? limit = null) =>
        GetAsync<TraceSequencesResult>(WithQuery("/control/traces/sequences",
            ("project", project),
            ("tool", tool),
            ("outcome", outcome),
            ("minDurationMs", minDurationMs?.ToString(CultureInfo.InvariantCulture)),
            ("maxDurationMs", maxDurationMs?.ToString(CultureInfo.InvariantCulture)),
            ("limit", limit?.ToString(CultureInfo.InvariantCulture))));

    /// <summary><c>GET /control/traces/failures</c> - its own endpoint, deliberately: failures are
    /// never to be filtered client-side out of the sequences list.</summary>
    public Task<FailedCallsResult> TracesFailuresAsync(string? project = null, int? limit = null) =>
        GetAsync<FailedCallsResult>(WithQuery("/control/traces/failures",
            ("project", project),
            ("limit", limit?.ToString(CultureInfo.InvariantCulture))));

    /// <summary><c>GET /control/traces/slow</c> - likewise its own endpoint.</summary>
    public Task<SlowToolsResult> TracesSlowAsync(string? project = null, int? limit = null) =>
        GetAsync<SlowToolsResult>(WithQuery("/control/traces/slow",
            ("project", project),
            ("limit", limit?.ToString(CultureInfo.InvariantCulture))));

    /// <summary>
    /// <c>GET /control/traces/{traceId}</c> - one call's full span detail.
    ///
    /// "sequences", "slow" and "failures" can never be mistaken for a trace id: ASP.NET Core prefers
    /// a literal segment over a route parameter at the same position, regardless of registration
    /// order, so the three routes above always win. See ControlListener's own comment.
    /// </summary>
    public Task<TraceDetailResult> TraceDetailAsync(string traceId, string? project = null) =>
        GetAsync<TraceDetailResult>(WithQuery(
            $"/control/traces/{Uri.EscapeDataString(traceId)}", ("project", project)));

    /// <summary><c>GET /control/memory</c> - authored documents and the proposal queue together, in
    /// the one round trip the endpoint provides.</summary>
    public Task<MemoryResult> MemoryAsync(string? project = null) =>
        GetAsync<MemoryResult>(WithQuery("/control/memory", ("project", project)));

    /// <summary>
    /// <c>GET /control/memory/document</c> - one document's complete raw text.
    ///
    /// <paramref name="name"/> is a QUERY parameter, not a route segment - see MemoryEndpoint's own
    /// doc comment for why. A document that does not exist answers a server error carrying its own
    /// explanation ("'{name}' does not exist yet."), which callers render verbatim.
    /// </summary>
    public Task<MemoryDocumentResult> MemoryDocumentAsync(string name, string? project = null) =>
        GetAsync<MemoryDocumentResult>(WithQuery("/control/memory/document",
            ("project", project), ("name", name)));

    /// <summary>
    /// <c>POST /control/memory/document</c> - OVERWRITES the named document.
    ///
    /// There is no merge and no version history: the core writes atomically over whatever was there.
    /// Memory is authored and irreplaceable, unlike the derived databases, so a caller must not
    /// reach this without an explicit confirmation - see MemoryViewModel's own gate.
    /// </summary>
    public Task<ActionResult> WriteMemoryDocumentAsync(string name, string content, string? project = null) =>
        PostAsync<WriteMemoryDocumentRequest, ActionResult>(
            WithQuery("/control/memory/document", ("project", project), ("name", name)),
            new WriteMemoryDocumentRequest { Content = content });

    /// <summary><c>POST /control/memory/proposals/accept</c>. Never destructive: accepting only ever
    /// APPENDS to the target document, creating it if it is missing.</summary>
    public Task<ActionResult> AcceptMemoryProposalAsync(string fileName, string? project = null) =>
        PostAsync<ActionResult>(WithQuery("/control/memory/proposals/accept",
            ("project", project), ("fileName", fileName)));

    /// <summary><c>POST /control/memory/proposals/defer</c>. Pure bookkeeping - never deletes, never
    /// writes an authored document.</summary>
    public Task<ActionResult> DeferMemoryProposalAsync(string fileName, string? project = null) =>
        PostAsync<ActionResult>(WithQuery("/control/memory/proposals/defer",
            ("project", project), ("fileName", fileName)));

    /// <summary>
    /// <c>POST /control/memory/proposals/dismiss</c> - DELETES the proposal file.
    ///
    /// <paramref name="confirm"/> defaults to false server-side and the endpoint refuses without it,
    /// so this parameter is the server's own gate rather than a client-side nicety. It is deliberately
    /// not defaulted to true here: a caller has to say so.
    /// </summary>
    public Task<ActionResult> DismissMemoryProposalAsync(string fileName, bool confirm, string? project = null) =>
        PostAsync<ActionResult>(WithQuery("/control/memory/proposals/dismiss",
            ("project", project),
            ("fileName", fileName),
            ("confirm", confirm ? "true" : "false")));

    // MARK: - Request plumbing

    /// <summary>Appends only the parameters that have a value. An absent parameter and an empty one
    /// are different things to this API - absent means "no filter", so a null is dropped rather than
    /// sent as an empty string.</summary>
    static string WithQuery(string path, params (string Name, string? Value)[] parameters)
    {
        var present = parameters
            .Where(p => !string.IsNullOrEmpty(p.Value))
            .Select(p => $"{Uri.EscapeDataString(p.Name)}={Uri.EscapeDataString(p.Value!)}")
            .ToArray();

        return present.Length == 0 ? path : $"{path}?{string.Join("&", present)}";
    }

    async Task<TResponse> GetAsync<TResponse>(string path)
    {
        using var request = new HttpRequestMessage(HttpMethod.Get, path);
        return await SendAsync<TResponse>(request);
    }

    async Task<TResponse> PostAsync<TResponse>(string path)
    {
        using var request = new HttpRequestMessage(HttpMethod.Post, path);
        return await SendAsync<TResponse>(request);
    }

    async Task<TResponse> PostAsync<TRequest, TResponse>(string path, TRequest body)
    {
        using var request = new HttpRequestMessage(HttpMethod.Post, path)
        {
            // The DTO's own [JsonPropertyName] attributes carry the wire names, exactly as they do
            // when decoding - no serializer options here for the same reason SendAsync uses none
            // when reading: anything configured on one side and not the other is drift waiting to
            // happen, and the golden fixtures pin the shape either way.
            Content = JsonContent.Create(body),
        };

        return await SendAsync<TResponse>(request);
    }

    /// <summary>The one call site every request above goes through - maps a transport failure, a
    /// non-2xx status, or a malformed body onto one of the four <see cref="ControlClientError"/>
    /// cases, mirroring the Swift original's own <c>send&lt;Response&gt;</c> case for case.</summary>
    async Task<TResponse> SendAsync<TResponse>(HttpRequestMessage request)
    {
        HttpResponseMessage response;
        try
        {
            response = await _http.SendAsync(request);
        }
        catch (Exception ex)
        {
            // HttpClient.SendAsync's documented failure mode is HttpRequestException (the core is
            // not running, this connection's port is stale, and so on); anything else still needs
            // to surface as a transport failure rather than crash - the request never got a
            // response to check the status of either way, matching the Swift original's own
            // `catch let error as URLError` plus its unconditional catch-all fallback.
            throw new ControlClientException(ControlClientError.Transport, ex.Message, ex);
        }

        var body = await response.Content.ReadAsStringAsync();

        if (!response.IsSuccessStatusCode)
        {
            if (response.StatusCode == HttpStatusCode.Unauthorized)
            {
                throw new ControlClientException(
                    ControlClientError.StaleToken, "The token this client presented was rejected.");
            }

            // The server's own "error" field, never text invented client-side - see
            // ControlErrorBody below and the Swift original's identical ControlErrorBody. A body
            // that fails to parse (or carries no "error" field) falls back to naming the raw
            // status code rather than throwing a second, unrelated decoding error out of an error
            // path.
            var message = TryReadErrorMessage(body) ?? $"Request failed with status {(int)response.StatusCode}.";
            throw new ControlClientException(
                ControlClientError.Server, message, innerException: null, statusCode: (int)response.StatusCode);
        }

        try
        {
            return JsonSerializer.Deserialize<TResponse>(body)
                ?? throw new ControlClientException(ControlClientError.Decoding, "Response body decoded to null.");
        }
        catch (JsonException ex)
        {
            throw new ControlClientException(ControlClientError.Decoding, ex.Message, ex);
        }
    }

    static string? TryReadErrorMessage(string body)
    {
        try
        {
            return JsonSerializer.Deserialize<ControlErrorBody>(body)?.Error;
        }
        catch (JsonException)
        {
            return null;
        }
    }
}

/// <summary>The <c>{"error": "..."}</c> shape every non-2xx Control response body carries - see
/// <c>ControlAuth.UseControlTokenAuth</c>, <c>ProjectsEndpoint.Remove</c>,
/// <c>EditorsEndpoint.ReleaseAsync</c>, and every other error path in the control API. Not a
/// public DTO: it exists only so <see cref="ControlClient"/> can read the server's own message
/// for <see cref="ControlClientError.Server"/> rather than inventing one - the .NET twin of the
/// Swift original's own private <c>ControlErrorBody</c>.</summary>
sealed record ControlErrorBody
{
    [JsonPropertyName("error")] public string? Error { get; init; }
}
